src/msca/metrics/main.py

```diff
@@ -0,0 +1,249 @@
+            Observed values
+        pred_values : np.ndarray
+            Predicted values
+        weight_values : np.ndarray
+            Sample weights
+
+        Returns
+        -------
+        float
+            Calculated metric value
+        """
+        match self:
+            case Metric.ROOT_MEAN_SQUARED_ERROR:
+                mse_value = metrics.mean_squared_error(
+                    y_true=obs_values,
+                    y_pred=pred_values,
+                    sample_weight=weight_values,
+                )
+                return np.sqrt(mse_value)
+            case Metric.MEAN_ABSOLUTE_ERROR:
+                return metrics.mean_absolute_error(
+                    y_true=obs_values,
+                    y_pred=pred_values,
+                    sample_weight=weight_values,
+                )
+            case Metric.MEAN_SQUARED_ERROR:
+                return metrics.mean_squared_error(
+                    y_true=obs_values,
+                    y_pred=pred_values,
+                    sample_weight=weight_values,
+                )
+            case Metric.MEAN_ABSOLUTE_PERCENTAGE_ERROR:
+                return metrics.mean_absolute_percentage_error(
+                    y_true=obs_values,
+                    y_pred=pred_values,
+                    sample_weight=weight_values,
+                )
+            case Metric.MEDIAN_ABSOLUTE_ERROR:
+                return metrics.median_absolute_error(
+                    y_true=obs_values,
+                    y_pred=pred_values,
+                    sample_weight=weight_values,
+                )
+            case _:
+                raise ValueError(f"Unsupported metric type: {self}")
+
+    def _eval_grouped(
+        self,
+        data: pd.DataFrame,
+        obs: str,
+        pred_ref: str,
+        weights: str,
+        groupby: list[str],
+        pred_alt: str | None = None,
+    ) -> pd.DataFrame:
+        """
+        Calculate error metrics or skill scores for each group in the DataFrame.
+
+        Parameters
+        ----------
+        data : pd.DataFrame
+            Input DataFrame
+        obs : str
+            Observed values column name
+        pred_ref : str
+            Reference predicted values column name
+        weights : str
+            Weights column name
+        groupby : list[str]
+            Grouping column names
+        pred_alt : str, optional
+            Alternative predicted values column name. If provided, calculates skill scores.
+
+        Returns
+        -------
+        pd.DataFrame
+            DataFrame with groupby columns and calculated metric/skill column
+        """
+        df = data.copy()
+
+        if pred_alt is None:
+            result_column_name = f"{pred_ref}_{self.value}"
+        else:
+            result_column_name = f"{pred_alt}_{self.value}_skill"
+
+        grouped_results = (
+            df.groupby(groupby)
+            .apply(
+                self._calculate_group_result,
```

Comment:
After change `_eval_single`, you can just directly pass in `self._eval_single` here.